pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,432 @@
+// Supporting more reference dates will imply more logic related to dates.
+const DAY_1: u32 = 1;
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time, in secs
+	Fixed {
+		/// Secs when maturity ends
+		date: Seconds,
+		/// Extension in secs, without special permissions
+		extension: Seconds,
+	},
+}
+
+impl Maturity {
+	pub fn fixed(date: Seconds) -> Self {
+		Self::Fixed { date, extension: 0 }
+	}
+
+	pub fn date(&self) -> Seconds {
+		match self {
+			Maturity::Fixed { date, .. } => *date,
+		}
+	}
+
+	pub fn is_valid(&self, now: Seconds) -> bool {
+		match self {
+			Maturity::Fixed { date, .. } => *date > now,
+		}
+	}
+
+	pub fn extends(&mut self, value: Seconds) -> Result<(), ArithmeticError> {
+		match self {
+			Maturity::Fixed { date, extension } => {
+				date.ensure_add_assign(value)?;
+				extension.ensure_sub_assign(value)
+			}
+		}
+	}
+}
+
+/// Interest payment periods
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum InterestPayments {
+	/// All interest is expected to be paid at the maturity date
+	None,
+
+	/// Interest is expected to be paid monthly
+	/// The associated value correspond to the paydown day in the month,
+	/// from 1-31.
+	/// The day will be adjusted to the month.
+	///
+	/// NOTE: Only day 1 is supported by now
+	Monthly(u8),
+}
+
+/// Specify the paydown schedules of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PayDownSchedule {
+	/// The entire borrowed amount is expected to be paid back at the maturity
+	/// date
+	None,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct CashflowPayment<Balance> {
+	pub when: Seconds,
+	pub principal: Balance,
+	pub interest: Balance,
+}
+
+/// Specify the repayment schedule of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct RepaymentSchedule {
+	/// Expected repayments date for remaining debt
+	pub maturity: Maturity,
+
+	/// Period at which interest is paid
+	pub interest_payments: InterestPayments,
+
+	/// How much of the initially borrowed amount is paid back during interest
+	/// payments
+	pub pay_down_schedule: PayDownSchedule,
+}
+
+impl RepaymentSchedule {
+	pub fn is_valid(&self, now: Seconds) -> Result<bool, DispatchError> {
+		match self.interest_payments {
+			InterestPayments::None => (),
+			InterestPayments::Monthly(_) => {
+				let start = date::from_seconds(now)?;
+				let end = date::from_seconds(self.maturity.date())?;
+
+				// We want to avoid creating a loan with a cashflow consuming a lot of computing
+				// time Maximum 40 years, which means a cashflow list of 40 * 12 elements
+				if end.year() - start.year() > 40 {
+					return Ok(false);
+				}
```

Comment:
We need to set a maximum to avoid computing big cashflows that could block the chain.

40 years is ok @hieronx? I do not know loans higher than that.

Reply:
Yes definitely 40 years is ok :)